### DISCUSSION TITLE
Generated images are blurry after using train_dreambooth_lora_sdxl_advanced.py to train a lora model

### DISCUSSION THREAD

**Discussion by joey0922** [huggingface/diffusers#8942] [Q&A] [answered]
I used script [train_dreambooth_lora_sdxl_advanced.py](https://github.com/huggingface/diffusers/blob/main/examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py)  from offical examples of diffusers to train a dreambooth and ti lora model by my own dataset, but images generated by this lora model are always blurry. I just changed the parameter max_train_steps without anything else changed and whatever how many steps I trained, the generated images are blurry. I believe my dataset to train is good with high quality and resolution. Can anybody help me out of this?

**Comment by tolgacangoz** [comment]
Could you examine [its blog post](https://huggingface.co/blog/sdxl_lora_advanced_script)?

**Reply by joey0922** [reply]
yes, I read it and tried, but the result is still the same.

**Comment by asomoza** [comment]
Hi, if you want help you will need to provide more info, we can't just guess what's your problem is. We have tested this script a lot of times and it works, so it can be literally anything.

First test it with a known dataset, so you can test it with just the [dog](https://huggingface.co/datasets/diffusers/dog-example) example and see if it generates good results. If it works, the problem is your dataset.

if not, you will need to provide more information so we can help you, ideally it should be:

- your environment, including the diffusers version
- the model you're using.
- your dataset.
- your train command.

Without these, it's impossible for us to know what the problem is.

**Reply by joey0922** [reply]
Thanks for your reply, I fixed it myself.

**Reply by tolgacangoz** [reply]
How? I wonder what the reason was. Is it possible for you to share your experience with us?

**Reply by joey0922** [reply, marked as answer]
I am not quite sure, I just raise the value of repeats, and modify a suitable value for max train steps then it worked. But I found though I set the value of repeats back to 1, it still worked. I have no idea why it happened. I didn't change anything else.

**Reply by joey0922** [reply]
@asomoza @tolgacangoz I want to ask another question. How to train multiple concepts at the same time by this script? What I know for now is that I can set multiple values to parameter "token_abstraction" according to specific concepts. For example, if I just want train a model to generate different car such as Benz and Audi, I can set a string "Benz,Audi" to token_abstraction, and I also have captioned images for train. According to the car in image, I used different identifier in the prompt of the image. That is to say a image of Benz car, I captioned it starting with "a photo of Benz car". Is it correct? What else I need to do for multi-concept training? It seems that the parameters "instance_prompt" and "validation_prompt" are not able to cover all identifiers. It is invalid to set more than one prompts to them?

**Reply by asomoza** [reply]
I think that script is not made for multiple concepts. The main target of dreambooth is to train a single concept really fast with a very small dataset, so the best traditional answer to your question is that you should probably train multiple loras and then merge them and see if they work together. Specially if you don't want to put too much effort in it.

My personal answer and my experience is that I never train with a TI (because I think it makes the loading more complicated and the benefits are not clear for me) and I don't train on special tokens. 

So what I do is to just generate a dataset with normal text descriptions and let the model learn, but for this to work you will need a bigger dataset and better descriptions, because the model needs to learn the difference between the concepts of each image and this will also take longer to train than a single concept DB LoRA. 

This is more similar to traditional LoRA training but in my experience it works better/faster if you have similar concepts, like for example some cars, or a person with different attributes like long hair, short hair, glasses, etc.

I think most of the popular loras are trained like this too, you can test this because the trigger words are not something like `sks` or some special token, they train with something normal like "pixel art", "papercut" or depending on what you're training you can even do it without a trigger word.

**Reply by joey0922** [reply]
Thanks for you valuable advice. And is there any experience or material you can share with me? I've thought of training normal lora models, but I don't have enough images for train. If I wanna train a lora model by a dataset with normal text descriptions, how many images in dataset do you think is good for this job?

**Reply by asomoza** [reply]
>  is there any experience or material you can share with me

Sadly there isn't that much material too read about this because even a difference in a couple of images will affect the training a lot, specially with dreambooth. So there isn't a standard "recipe" to use, finetuning and training LoRAs (with image generation), IMO is learned a lot more by experimenting than by reading.

> If I wanna train a lora model by a dataset with normal text descriptions, how many images in dataset do you think is good for this job?

It depends a lot in what you want so as before there isn't a number I can't give you about this. 

Training specific persons doesn't need that much quantity of images but they need a really good description and quality, one bad image will make the training useless the same if you miss one description or put a wrong word. The hard part here with the descriptions is to know which is considered the "person" you want to train and what isn't.

Also people normally use really high rank numbers and learning rates which makes the training faster and to "seem" better but in reality they overfit the model and mostly the person trained can do just a couple of fixed poses and you won't be able to merge different loras with it.

Styles are more forgiving in the sense of quality and descriptions but you will need a lot more images for the model to learn it, I'm taking hundreds or even thousand of them, depending on the style and if the model knows something similar from before. For styles you can just use a good VLLM to generate the descriptions.

We're going to do sometime soon a blog post about this, also personally I really want to do a guide but I need a dataset and this is really hard to do because I need a public domain (or truly open source one) and safe one, and most of the models are already trained with those.

**Reply by linoytsaban** [reply]
+100 to @asomoza insightful comments 🙌🏻

First to add to this comment: 

> Also people normally use really high rank numbers and learning rates which makes the training faster and to "seem" better but in reality they overfit the model and mostly the person trained can do just a couple of fixed poses and you won't be able to merge different loras with it.

I very much agree with that. Another thing that is important to keep in mind that I think sometimes gets overlooked is that training is also **extremely** affected by how much the concept you're training on is already familiar to the base model. Some LoRAs get amazing results, but when you prompt the base model for the concept you get similar quality results (i.e. the base model already knew that concept pretty well) so you can't learn much about the dataset and parameter choice used for training, as the model didnt need much learning to begin with. 

 Also to add on my end re: training with TI & multi concepts LoRAs- 
- I do train with TI mostly. When we add text encoder training (compared to regular training of LoRA weights for the Unet only) we do it to better capture our trained concept and hence if we alter the weights of the text encoder too, we're not only changing the way the Unet perceives the concept, but also the way it's represented in the text-encoder space. With TI we're also changing the way the concept is represented in the text-encoder space, but from a different angle - instead of modifying the weights of the text-encoders, we introduce new tokens and during the training process modify the text-embeddings associated with those, the idea is that by initializing new tokens we're less affected by priors because there isn't a concept already associated with those tokens. In my experience using TI led to better training of the **unet** and in many times results were better/equivalent to text encoder training except the only weights changed here are the unet's.

- multi concept training is a difficult task, but one thing that has been done (not super successfully) in the past is to train a LoRA one concept 1, and then re-train on concept 2, while using prior preservation loss with concept 1 to "enforce" the model not to "forget" it. You can also try that and use TI for both with different tokens

**Reply by asomoza** [reply]
Thanks @linoytsaban for explaining the reasons of the TI training, it really helps to understand it better.

And yeah, I pretty much forgot to tell that I don't use TI because I always train the text encoders and I also load the loras with different scales for the unet blocks and text encoders so that makes a difference too. 

They're just different ways of training and doing inference which is a good thing to have.
